### DISCUSSION TITLE
Is there a existing way to log each operation when running upgrade?

### DISCUSSION THREAD

**Discussion by notatallshaw-gts** [sqlalchemy/alembic#1344] [Usage Questions] [answered]
**Describe the use case**

We're having an issue where some of our revisions are _intermittently_ failing, i.e. running `alembic upgrade heads` against a clean database will usually work but sometimes 2 particular revisions will fail. 

The two revisions contain many `op.{method}` statements to create tables and add foreign keys. And while the error implies that there is an issue creating the table, we don't want to rule out it's actually creating a foreign key after the table is created and the error is misleading.

**Databases / Backends / Drivers targeted**

MySQL/pymysql

**Additional context**

I appreciate this is more of a support question than an issue, but I just assumed there would be an option to log each operations, not just each revision, but looking through the code I don't spot anything. I also searched exiting issues but nothing stood out to me.

**Comment by zzzeek** [comment, marked as answer]
SQLAlchemy logging in general is controlled by a Python logger called `sqlalchemy.engine`, which when set to INFO or higher will log every SQL statement and transaction control command emitted.  General info on this logger is at https://docs.sqlalchemy.org/en/20/core/engines.html#configuring-logging .  Within Alembic, if you are working with a default `alembic init`, you would go into the provided `alembic.ini` and set the SQLAlchemy logger as:

```
[logger_sqlalchemy]
level = INFO
handlers =
qualname = sqlalchemy.engine

```

**Reply by notatallshaw-gts** [reply]
Thanks, this provides the required information.
